test/e2e/hairpin.go

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"fmt"
+	"time"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/util/intstr"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = framework.KubeDescribe("Hairpin Networking", func() {
+	f := framework.NewDefaultFramework("hairpintest")
+
+	var svcname = "hairpintest"
+
+	It("should be able to contact a service served by the same pod", func() {
```

Comment:
I think we already test this via networking e2es because we create a Service over some pods, then exec into one and verify that it can reach all endpoints (including itself) of the Service through clusterIP https://github.com/kubernetes/kubernetes/blob/master/test/e2e/networking.go#L89

Though a more isolated test definitely can't hurt, but can you just add it to test/e2e/networking.go or servies.go?